File: moqt/internal/message/track_info.go
package message

import (
	"io"
)

/*
 *	TRACK_INFO Message {
 *	  Message Length (i)
 *	  Publisher Priority (8)
 *	  Publisher Ordered (8)
 *	  Publisher Max Latency (i)
 *	  Timescale (i)
 *	}
 *
 * Every field is fixed for the lifetime of the Track.
 * Timescale MUST be non-zero.
 */
type TrackInfoMessage struct {
	PublisherPriority   uint8
	PublisherOrdered    uint8
	PublisherMaxLatency uint64
	Timescale           uint64
}

func (tim TrackInfoMessage) Len() int {
	var l int

	l += 1 // PublisherPriority (uint8)
	l += 1 // PublisherOrdered (uint8)
	l += VarintLen(tim.PublisherMaxLatency)
	l += VarintLen(tim.Timescale)

	return l
}

func (tim TrackInfoMessage) Encode(w io.Writer) error {
	msgLen := tim.Len()
	b := make([]byte, 0, msgLen+VarintLen(uint64(msgLen)))

	b, _ = WriteMessageLength(b, uint64(msgLen))
	b = append(b, tim.PublisherPriority)
	b = append(b, tim.PublisherOrdered)
	b, _ = WriteVarint(b, tim.PublisherMaxLatency)
	b, _ = WriteVarint(b, tim.Timescale)

	_, err := w.Write(b)

	return err
}

func (tim *TrackInfoMessage) Decode(src io.Reader) error {
	size, err := ReadMessageLength(src)
	if err != nil {
		return err
	}

	if size > MaxMessageSize {
		return ErrMessageTooLarge
	}

	b := make([]byte, size)

	_, err = io.ReadFull(src, b)
	if err != nil {
		return err
	}

	if len(b) < 2 {
		return ErrMessageTooShort
	}
	tim.PublisherPriority = b[0]
	tim.PublisherOrdered = b[1]
	b = b[2:]

	num, n, err := ReadVarint(b)
	if err != nil {
		return err
	}
	tim.PublisherMaxLatency = num
	b = b[n:]

	num, n, err = ReadVarint(b)
	if err != nil {
		return err
	}
	tim.Timescale = num
	b = b[n:]

	if len(b) != 0 {
		return ErrMessageTooShort
	}

	return nil
}

File: moq-web/src/internal/message/track_info.ts
import type { Reader, Writer } from "@okdaichi/golikejs/io";
import { MessageDecoder, MessageEncoder, readFull, readVarint } from "./message.ts";

export interface TrackInfoMessageInit {
	publisherPriority?: number;
	publisherOrdered?: number;
	publisherMaxLatency?: number;
	timescale?: number;
}

/**
 * TRACK_INFO message, sent by the publisher as the sole response on a
 * Track Stream. Every field is fixed for the lifetime of the track;
 * Timescale MUST be non-zero.
 *
 * ```text
 * TRACK_INFO Message {
 *   Message Length (i)
 *   Publisher Priority (8)
 *   Publisher Ordered (8)
 *   Publisher Max Latency (i)
 *   Timescale (i)
 * }
 * ```
 */
export class TrackInfoMessage {
	publisherPriority: number;
	publisherOrdered: number;
	publisherMaxLatency: number;
	timescale: number;

	constructor(init: TrackInfoMessageInit = {}) {
		this.publisherPriority = init.publisherPriority ?? 0;
		this.publisherOrdered = init.publisherOrdered ?? 0;
		this.publisherMaxLatency = init.publisherMaxLatency ?? 0;
		this.timescale = init.timescale ?? 0;
	}

	/**
	 * Encodes the message to the writer.
	 */
	async encode(w: Writer): Promise<Error | undefined> {
		const e = new MessageEncoder();
		e.uint8(this.publisherPriority);
		e.uint8(this.publisherOrdered);
		e.varint(this.publisherMaxLatency);
		e.varint(this.timescale);
		const [, err] = await w.write(e.frame());
		return err;
	}

	/**
	 * Decodes the message from the reader.
	 */
	async decode(r: Reader): Promise<Error | undefined> {
		const [msgLen, , err1] = await readVarint(r);
		if (err1) return err1;

		const buf = new Uint8Array(msgLen);
		const [, err2] = await readFull(r, buf);
		if (err2) return err2;

		const d = new MessageDecoder(buf);
		this.publisherPriority = d.uint8();
		this.publisherOrdered = d.uint8();
		this.publisherMaxLatency = d.varint();
		this.timescale = d.varint();

		return undefined;
	}
}
